crates/cargo-util/src/registry.rs

```diff
@@ -11,11 +11,20 @@ pub fn make_dep_path(dep_name: &str, prefix_only: bool) -> String {
     } else {
         ("/", dep_name)
     };
-    match dep_name.len() {
+    match dep_name.chars().take(4).count() {
         1 => format!("1{}{}", slash, name),
         2 => format!("2{}{}", slash, name),
-        3 => format!("3/{}{}{}", &dep_name[..1], slash, name),
-        _ => format!("{}/{}{}{}", &dep_name[0..2], &dep_name[2..4], slash, name),
+        3 => {
+            let first_symbol = dep_name.chars().take(1).collect::<String>();
+
+            format!("3/{}{}{}", first_symbol, slash, name)
+        }
+        _ => {
+            let first_symbol = dep_name.chars().take(2).collect::<String>();
+            let second_symbol = dep_name.chars().skip(2).take(2).collect::<String>();
+
+            format!("{}/{}{}{}", first_symbol, second_symbol, slash, name)
+        }
```

Comment:
Sorry for my premature optimization, could we avoid allocations once again? This function will be called several times in a build. Also, we should assert empty string as well since we don't want to create a ill-formatted path.

```suggestion
        l => {
            assert!(l > 0, "length of a dependency name must not be zero");
            let mut it = dep_name.chars();
            let (f0, f1) = (it.next().unwrap(), it.next().unwrap());
            let (s0, s1) = (it.next().unwrap(), it.next().unwrap());
            format!("{f0}{f1}/{s0}{s1}{slash}{name}")
        }
```

Reply:
I am totally fine with you not accepting this ugly optimization, though I'd prefer to keep an assertion here (though it's unlikely to happen).